server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/managable/cassandra/EnqueuedMailsDao.java

```diff
@@ -0,0 +1,391 @@
+            .setString(STATE, mail.getState())
+            .setString(SENDER, Optional.ofNullable(mail.getSender())
+                .map(MailAddress::asString)
+                .orElse(null))
+            .setList(RECIPIENTS, asStringList(mail.getRecipients()))
+            .setString(ERROR_MESSAGE, mail.getErrorMessage())
+            .setString(REMOTE_ADDR, mail.getRemoteAddr())
+            .setString(REMOTE_HOST, mail.getRemoteHost())
+            .setTimestamp(LAST_UPDATED, mail.getLastUpdated())
+            .setMap(ATTRIBUTES, toRawAttributeMap(mail))
+            .setMap(PER_RECIPIENT_SPECIFIC_HEADERS, toHeaderMap(cassandraTypesProvider, mail.getPerRecipientSpecificHeaders()))
+        );
+    }
+
+    public CompletableFuture<Stream<ManageableMailQueue.MailQueueItemView>> browse(
+        MailQueueName queueName, Date timeRangeStartFrom) {
+
+        return CompletableFutureUtil.allOf(
+            bucketAndSliceStream(timeRangeStartFrom)
+                .map(bucketAndSlice -> browse(queueName, bucketAndSlice.getBucketId(), bucketAndSlice.getSliceStartInstant())))
+            .thenApply(stream -> stream.flatMap(Function.identity()));
+    }
+
+    private CompletableFuture<Stream<ManageableMailQueue.MailQueueItemView>> browse(
+        MailQueueName queueName, Integer bucketId, Instant sliceStartInstant) {
+
+        CompletableFuture<Stream<CassandraManageableMailQueueItem>> enqueuedItemsFuture = executor.execute(
+            selectFrom.bind()
+                .setString(QUEUE_NAME, queueName.asString())
+                .setTimestamp(TIME_RANGE_START, Date.from(sliceStartInstant))
+                .setInt(BUCKET_ID, bucketId))
+            .thenApply(resultSet -> cassandraUtils.convertToStream(resultSet)
+                .map(EnqueuedMailsDaoUtil::toMailQueueItemView));
+
+        return enqueuedItemsFuture
+            .thenCompose(this::filterNonDeletedItem);
+    }
+
+    private CompletableFuture<Stream<ManageableMailQueue.MailQueueItemView>> filterNonDeletedItem(
+        Stream<CassandraManageableMailQueueItem> manageableMailQueueItemStream) {
+
+        return CompletableFutureUtil.allOf(
+            manageableMailQueueItemStream
+                .map(deletedMailsDao::filterNotDeleted))
+            .thenApply(this::filterPresentItem);
+    }
+
+    private Stream<ManageableMailQueue.MailQueueItemView> filterPresentItem(
+        Stream<Optional<CassandraManageableMailQueueItem>> afterCheckingDeletionStream) {
+
+        return afterCheckingDeletionStream
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .map(CassandraManageableMailQueueItem::getMailQueueItemView);
+    }
+
+    private Stream<BucketAndSlice> bucketAndSliceStream(Date timeRangeStartFrom) {
+        Stream<Instant> sliceStartStream = sliceStartStream(timeRangeStartFrom);
+        Stream<Integer> bucketStream = IntStream.range(0, manageableQueueConfiguration.getBucketCount()).boxed();
+
+        return bucketStream
```

Comment:
What you are doing here is a very common operation called 'cardinal product'.

I propose to have a method of Streamutils doing it:

```
Stream<R> cardinalProduct(Stream<A> s1, Stream<B> s2, BiFuction<A, B, R> combinator);
```

That would allow to instantly recognise what you are doing.